File: apps/tui/src/projects.test.ts
import { describe, expect, test } from "bun:test"

import {
  createProject,
  deleteProject,
  listProjects,
  refreshProject,
  type ProjectFetch,
} from "./projects"

const projectPayload = {
  id: "project-1",
  name: "Example",
  repositories: [
    {
      id: "repository-1",
      project_id: "project-1",
      local_path: "/tmp/example",
      current_branch: "main",
      head_sha: "abc123",
      dirty: false,
      created_at: "2026-07-31T00:00:00Z",
      updated_at: "2026-07-31T00:00:00Z",
    },
  ],
  created_at: "2026-07-31T00:00:00Z",
  updated_at: "2026-07-31T00:00:00Z",
}

describe("project API client", () => {
  test("lists projects", async () => {
    const fetcher: ProjectFetch = async () =>
      new Response(JSON.stringify({ data: [projectPayload] }), { status: 200 })

    const projects = await listProjects(fetcher)
    expect(projects).toHaveLength(1)
    expect(projects[0]?.repositories[0]?.head_sha).toBe("abc123")
  })

  test("creates a project with repository path", async () => {
    let requestBody = ""
    const fetcher: ProjectFetch = async (_input, init) => {
      requestBody = String(init?.body)
      return new Response(JSON.stringify({ data: projectPayload }), { status: 201 })
    }

    const project = await createProject("Example", "/tmp/example", fetcher)
    expect(project.id).toBe("project-1")
    expect(JSON.parse(requestBody)).toEqual({
      name: "Example",
      repository_path: "/tmp/example",
    })
  })

  test("surfaces API errors", async () => {
    const fetcher: ProjectFetch = async () =>
      new Response(JSON.stringify({ error: { message: "repository is already registered" } }), {
        status: 409,
      })

    await expect(createProject("Example", "/tmp/example", fetcher)).rejects.toThrow(
      "repository is already registered",
    )
  })

  test("deletes and refreshes a project", async () => {
    const requests: string[] = []
    const fetcher: ProjectFetch = async (input, init) => {
      requests.push(`${init?.method} ${String(input)}`)
      if (init?.method === "DELETE") {
        return new Response(null, { status: 204 })
      }
      return new Response(JSON.stringify({ data: projectPayload }), { status: 200 })
    }

    await deleteProject("project-1", fetcher)
    const refreshed = await refreshProject("project-1", fetcher)

    expect(refreshed.name).toBe("Example")
    expect(requests[0]).toContain("DELETE")
    expect(requests[1]).toContain("POST")
  })
})

File: internal/httpapi/projects_test.go
package httpapi

import (
	"context"
	"errors"
	"net/http"
	"net/http/httptest"
	"strings"
	"testing"
	"time"

	"github.com/livingdolls/orkoda-tui/internal/projects"
)

type fakeProjectRegistry struct {
	projects       []projects.Project
	createdName    string
	createdPath    string
	renamedID      string
	renamedName    string
	deletedID      string
	refreshedID    string
	operationError error
}

func (f *fakeProjectRegistry) Create(_ context.Context, name, repositoryPath string) (projects.Project, error) {
	if f.operationError != nil {
		return projects.Project{}, f.operationError
	}
	f.createdName = name
	f.createdPath = repositoryPath
	return f.projects[0], nil
}

func (f *fakeProjectRegistry) List(context.Context) ([]projects.Project, error) {
	if f.operationError != nil {
		return nil, f.operationError
	}
	return f.projects, nil
}

func (f *fakeProjectRegistry) Get(_ context.Context, projectID string) (projects.Project, error) {
	if f.operationError != nil {
		return projects.Project{}, f.operationError
	}
	for _, project := range f.projects {
		if project.ID == projectID {
			return project, nil
		}
	}
	return projects.Project{}, projects.ErrNotFound
}

func (f *fakeProjectRegistry) Rename(_ context.Context, projectID, name string) (projects.Project, error) {
	if f.operationError != nil {
		return projects.Project{}, f.operationError
	}
	f.renamedID = projectID
	f.renamedName = name
	project := f.projects[0]
	project.Name = name
	return project, nil
}

func (f *fakeProjectRegistry) Delete(_ context.Context, projectID string) error {
	if f.operationError != nil {
		return f.operationError
	}
	f.deletedID = projectID
	return nil
}

func (f *fakeProjectRegistry) Refresh(_ context.Context, projectID string) (projects.Project, error) {
	if f.operationError != nil {
		return projects.Project{}, f.operationError
	}
	f.refreshedID = projectID
	return f.projects[0], nil
}

func testProject() projects.Project {
	now := time.Unix(100, 0).UTC()
	return projects.Project{
		ID:   "project-1",
		Name: "Example",
		Repositories: []projects.RepositoryInfo{
			{
				ID:            "repository-1",
				ProjectID:     "project-1",
				LocalPath:     "/tmp/example",
				CurrentBranch: "main",
				HeadSHA:       "abc123",
				CreatedAt:     now,
				UpdatedAt:     now,
			},
		},
		CreatedAt: now,
		UpdatedAt: now,
	}
}

func TestCreateAndListProjects(t *testing.T) {
	registry := &fakeProjectRegistry{projects: []projects.Project{testProject()}}
	router := NewRouter("development", nil, registry)

	createRequest := httptest.NewRequest(http.MethodPost, "/api/v1/projects", strings.NewReader(`{
		"name":"Example",
		"repository_path":"/tmp/example"
	}`))
	createRequest.Header.Set("content-type", "application/json")
	createResponse := httptest.NewRecorder()
	router.ServeHTTP(createResponse, createRequest)
	if createResponse.Code != http.StatusCreated {
		t.Fatalf("create status = %d, body = %s", createResponse.Code, createResponse.Body.String())
	}
	if registry.createdName != "Example" || registry.createdPath != "/tmp/example" {
		t.Fatalf("create args name=%q path=%q", registry.createdName, registry.createdPath)
	}

	listRequest := httptest.NewRequest(http.MethodGet, "/api/v1/projects", nil)
	listResponse := httptest.NewRecorder()
	router.ServeHTTP(listResponse, listRequest)
	if listResponse.Code != http.StatusOK || !strings.Contains(listResponse.Body.String(), `"head_sha":"abc123"`) {
		t.Fatalf("list status = %d, body = %s", listResponse.Code, listResponse.Body.String())
	}
}

func TestProjectMutationEndpoints(t *testing.T) {
	registry := &fakeProjectRegistry{projects: []projects.Project{testProject()}}
	router := NewRouter("development", nil, registry)

	renameRequest := httptest.NewRequest(http.MethodPatch, "/api/v1/projects/project-1", strings.NewReader(`{"name":"Renamed"}`))
	renameRequest.Header.Set("content-type", "application/json")
	renameResponse := httptest.NewRecorder()
	router.ServeHTTP(renameResponse, renameRequest)
	if renameResponse.Code != http.StatusOK || registry.renamedName != "Renamed" {
		t.Fatalf("rename status=%d body=%s", renameResponse.Code, renameResponse.Body.String())
	}

	refreshRequest := httptest.NewRequest(http.MethodPost, "/api/v1/projects/project-1/refresh", nil)
	refreshResponse := httptest.NewRecorder()
	router.ServeHTTP(refreshResponse, refreshRequest)
	if refreshResponse.Code != http.StatusOK || registry.refreshedID != "project-1" {
		t.Fatalf("refresh status=%d body=%s", refreshResponse.Code, refreshResponse.Body.String())
	}

	deleteRequest := httptest.NewRequest(http.MethodDelete, "/api/v1/projects/project-1", nil)
	deleteResponse := httptest.NewRecorder()
	router.ServeHTTP(deleteResponse, deleteRequest)
	if deleteResponse.Code != http.StatusNoContent || registry.deletedID != "project-1" {
		t.Fatalf("delete status=%d body=%s", deleteResponse.Code, deleteResponse.Body.String())
	}
}

func TestProjectErrorsMapToHTTPStatus(t *testing.T) {
	tests := []struct {
		name   string
		err    error
		status int
	}{
		{name: "invalid", err: projects.ErrInvalidProject, status: http.StatusBadRequest},
		{name: "duplicate", err: projects.ErrRepositoryAlreadyRegistered, status: http.StatusConflict},
		{name: "not found", err: projects.ErrNotFound, status: http.StatusNotFound},
		{name: "internal", err: errors.New("database unavailable"), status: http.StatusInternalServerError},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			registry := &fakeProjectRegistry{projects: []projects.Project{testProject()}, operationError: test.err}
			router := NewRouter("development", nil, registry)
			request := httptest.NewRequest(http.MethodGet, "/api/v1/projects", nil)
			response := httptest.NewRecorder()
			router.ServeHTTP(response, request)
			if response.Code != test.status {
				t.Fatalf("status = %d, want %d, body = %s", response.Code, test.status, response.Body.String())
			}
		})
	}
}
